presto-main/src/main/java/com/facebook/presto/operator/scalar/ConcatFunction.java

```diff
@@ -0,0 +1,155 @@
+import static com.facebook.presto.byteCode.Access.PRIVATE;
+import static com.facebook.presto.byteCode.Access.PUBLIC;
+import static com.facebook.presto.byteCode.Access.STATIC;
+import static com.facebook.presto.byteCode.Access.a;
+import static com.facebook.presto.byteCode.Parameter.arg;
+import static com.facebook.presto.byteCode.ParameterizedType.type;
+import static com.facebook.presto.metadata.Signature.internalFunction;
+import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
+import static com.facebook.presto.sql.gen.CompilerUtils.defineClass;
+import static com.facebook.presto.util.Reflection.methodHandle;
+
+public final class ConcatFunction
+        extends ParametricScalar
+{
+    public static final ConcatFunction CONCAT = new ConcatFunction();
+    private static final Signature SIGNATURE = new Signature("concat", ImmutableList.of(), StandardTypes.VARCHAR, ImmutableList.of(StandardTypes.VARCHAR), true, false);
+
+    @Override
+    public Signature getSignature()
+    {
+        return SIGNATURE;
+    }
+
+    @Override
+    public boolean isHidden()
+    {
+        return false;
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return true;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "concatenates given strings";
+    }
+
+    @Override
+    public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        Class<?> clazz = generateConcat(arity);
+        MethodHandle methodHandle = methodHandle(clazz, "concat", Collections.nCopies(arity, Slice.class).toArray(new Class<?>[arity]));
+        List<Boolean> nullableParameters = ImmutableList.copyOf(Collections.nCopies(arity, false));
+
+        Signature specializedSignature = internalFunction(SIGNATURE.getName(), VARCHAR.getTypeSignature(), Collections.nCopies(arity, VARCHAR.getTypeSignature()));
+        return new FunctionInfo(specializedSignature, getDescription(), isHidden(), methodHandle, isDeterministic(), false, nullableParameters);
+    }
+
+    private static Class<?> generateConcat(int arity)
+    {
+        ClassDefinition definition = new ClassDefinition(
+                a(PUBLIC, FINAL),
+                CompilerUtils.makeClassName("Concat" + arity + "ScalarFunction"),
+                type(Object.class));
+
+        // Generate constructor
+        definition.declareDefaultConstructor(a(PRIVATE));
+
+        // Generate concat()
+        ImmutableList.Builder<Parameter> parameters = ImmutableList.builder();
+        for (int i = 0; i < arity; i++) {
+            parameters.add(arg("arg" + i, Slice.class));
+        }
+
+        MethodDefinition method = definition.declareMethod(a(PUBLIC, STATIC), "concat", type(Slice.class), parameters.build());
+        Scope scope = method.getScope();
+        Block body = method.getBody();
+
+        body.comment("put 0 as initial result varchar length on stack")
```

Comment:
Actually, rather than `add()`, handle overflow by adding and calling this method:
```java
public static int checkedAdd(int x, int y)
{
    try {
        return Math.addExact(x, y);
    }
    catch (ArithmeticException e) {
        throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Concatenated string is too large");
    }
}